### PR TITLE
Prepare for migration

### LINKED ISSUE #9290: ClusterPolicy updates are not idempotent due to lastModification
The presence of the lastModified value on a ClusterPolicy and ClusterPolicyBinding prevents non-mutating writes from being suppressed, thus leading to higher than necessary event propagation in the cluster and prevents migrations from being communicated correctly as being idempotent.

The mutation logic should check the difference between old and new before incrementing the modification timestamp.

Putting P1 for triage right now because every storage migration results in a mutation.

### PR DESCRIPTION
**Pull request by smarterclayton**
* Don't modify policies when they don't change. Fixes #9290
* Cleanup discovery a bit
  * Handle a cache race
  * Better message from server

### CODE REVIEW

**Comment by smarterclayton** [comment]
[test] @deads2k split at your request, addressed your comments (using the standard discovery thing)

File: pkg/authorization/registry/role/policybased/virtual_storage_test.go [hunk 1]
```diff
@@ -97,6 +97,9 @@ func TestUpdate(t *testing.T) {
 	ctx := kapi.WithUser(kapi.WithNamespace(kapi.NewContext(), "unittest"), &user.DefaultInfo{Name: "system:admin"})
 	realizedRoleObj, err := storage.Create(ctx, &authorizationapi.Role{
 		ObjectMeta: kapi.ObjectMeta{Name: "my-role"},
+		Rules: []authorizationapi.PolicyRule{
+			{Verbs: sets.NewString(authorizationapi.VerbAll)},
```
Comment (deads2k) [inline pkg/authorization/registry/role/policybased/virtual_storage_test.go]: why this change?

File: pkg/cmd/util/clientcmd/factory.go [hunk 2]
```diff
@@ -774,3 +775,51 @@ func (c *clientCache) ClientForVersion(version *unversioned.GroupVersion) (*clie
...
 	return client, nil
 }
+
+// FindAllCanonicalResources returns all resource names that map directly to their kind (Kind -> Resource -> Kind)
+// and are not subresources. This is the closest mapping possible from the client side to resources that can be
+// listed and updated. Note that this may return some virtual resources (like imagestreamtags) that can be otherwise
+// represented.
+// TODO: add a field to APIResources for "virtual" (or that points to the canonical resource).
+// TODO: fallback to the scheme when discovery is not possible.
+func FindAllCanonicalResources(f *Factory) ([]unversioned.GroupResource, error) {
```
Comment (deads2k) [inline pkg/cmd/util/clientcmd/factory.go]: The factory is a pain to mock.  Looks like you want a `RESTMapper` and a `DiscoveryClient`.  How about just accepting those.

Comment (deads2k) [inline pkg/cmd/util/clientcmd/factory.go]: > The factory is a pain to mock. Looks like you want a RESTMapper and a DiscoveryClient. How about just accepting those.

That would also make this easier to use inside the `Run` step, since it won't need a full `Factory`.  Oh, and I really want to refactor that type.

File: pkg/cmd/util/clientcmd/factory.go [hunk 3]
```diff
@@ -774,3 +775,51 @@ func (c *clientCache) ClientForVersion(version *unversioned.GroupVersion) (*clie
...
+			// lookup (KindsFor) does not error.
+			if mapping, err := m.RESTMapping(unversioned.GroupKind{Group: gv.Group, Kind: r.Kind}, gv.Version); err == nil {
+				if _, err := m.KindsFor(mapping.GroupVersionKind.GroupVersion().WithResource(mapping.Resource)); err == nil {
+					set[mapping.Resource] = unversioned.GroupResource{Group: gv.Group, Resource: mapping.Resource}
+				}
+			}
+		}
+	}
+	keys := sets.NewString()
+	for k := range set {
```
Comment (deads2k) [inline pkg/cmd/util/clientcmd/factory.go]: `sets.StringKeySet`

File: pkg/cmd/util/clientcmd/factory.go [hunk 4]
```diff
@@ -774,3 +775,51 @@ func (c *clientCache) ClientForVersion(version *unversioned.GroupVersion) (*clie
...
+// represented.
+// TODO: add a field to APIResources for "virtual" (or that points to the canonical resource).
+// TODO: fallback to the scheme when discovery is not possible.
+func FindAllCanonicalResources(f *Factory) ([]unversioned.GroupResource, error) {
+	m, _ := f.Object(false)
+	oclient, _, err := f.Clients()
+	if err != nil {
+		return nil, err
+	}
+	set := make(map[string]unversioned.GroupResource)
```
Comment (deads2k) [inline pkg/cmd/util/clientcmd/factory.go]: Why not just `map[unversioned.GroupResource]bool`, then iterate over keys below?

File: pkg/authorization/registry/role/policybased/virtual_storage_test.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/authorization/registry/role/policybased/virtual_storage_test.go]: The test wasn't actually changing anything before.

File: pkg/cmd/util/clientcmd/factory.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/cmd/util/clientcmd/factory.go]: Not sure what you mean.

File: pkg/cmd/util/clientcmd/factory.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/cmd/util/clientcmd/factory.go]: At first I didn't want to deal with group resources, but I guess I'm going to have to.

File: pkg/cmd/util/clientcmd/factory.go (hunk 3 above)
Comment (deads2k) [inline pkg/cmd/util/clientcmd/factory.go]: > Not sure what you mean

`keys = sets.StringKeySet(set)`, though as I comment above, I think you can just restructure that map.

File: pkg/cmd/util/clientcmd/factory.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/cmd/util/clientcmd/factory.go]: Actually, I don't want cohabitation resources right now for this call.

File: pkg/cmd/util/clientcmd/factory.go [hunk 5]
```diff
@@ -774,3 +776,46 @@ func (c *clientCache) ClientForVersion(version *unversioned.GroupVersion) (*clie
...
 }
+
+// FindAllCanonicalResources returns all resource names that map directly to their kind (Kind -> Resource -> Kind)
+// and are not subresources. This is the closest mapping possible from the client side to resources that can be
+// listed and updated. Note that this may return some virtual resources (like imagestreamtags) that can be otherwise
+// represented. It collapses resources in different groups with the same name into a single value.
+// TODO: add a field to APIResources for "virtual" (or that points to the canonical resource).
+// TODO: fallback to the scheme when discovery is not possible.
+func FindAllCanonicalResources(d discovery.Discovery, m meta.RESTMapper) ([]unversioned.GroupResource, error) {
+	set := make(map[string]unversioned.GroupResource)
```
Comment (deads2k) [inline pkg/cmd/util/clientcmd/factory.go]: So if I'm reading this correctly, you're using this in a way that will fail if I get a `foo.group1` and `foo.group2`.  Those would have a single entry in the map, but they should have two.

Comment (smarterclayton) [inline pkg/cmd/util/clientcmd/factory.go]: I don't think they should (it's not the point of this method).  The
heuristic is resources with the same name share storage, and are henceforth
only migrated once.

On Jun 17, 2016, at 8:12 AM, David Eads <notifications@github.com> wrote:

In pkg/cmd/util/clientcmd/factory.go
<https://github.com/openshift/origin/pull/9380#discussion_r67500623>:

> @@ -774,3 +776,46 @@ func (c *clientCache) ClientForVersion(version *unversioned.GroupVersion) (*clie
>  	c.clients[config.GroupVersion.String()] = client
>  	return client, nil
>  }
> +
> +// FindAllCanonicalResources returns all resource names that map directly to their kind (Kind -> Resource -> Kind)
> +// and are not subresources. This is the closest mapping possible from the client side to resources that can be
> +// listed and updated. Note that this may return some virtual resources (like imagestreamtags) that can be otherwise
> +// represented. It collapses resources in different groups with the same name into a single value.
> +// TODO: add a field to APIResources for "virtual" (or that points to the canonical resource).
> +// TODO: fallback to the scheme when discovery is not possible.
> +func FindAllCanonicalResources(d discovery.Discovery, m meta.RESTMapper) ([]unversioned.GroupResource, error) {
> +	set := make(map[string]unversioned.GroupResource)

So if I'm reading this correctly, you're using this in a way that will fail
if I get a foo.group1 and foo.group2. Those would have a single entry in
the map, but they should have two.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/9380/files/fdc37b55eefd4f22bd8368bec25c51eb95c92009#r67500623>,
or mute the thread
<https://github.com/notifications/unsubscribe/ABG_p1WyVDmMM0Qd5MztOYGEpRTGIPM_ks5qMo80gaJpZM4I3dIQ>
.

Comment (deads2k) [inline pkg/cmd/util/clientcmd/factory.go]: > I don't think they should (it's not the point of this method).  The
> heuristic is resources with the same name share storage, and are henceforth
> only migrated once.

`roles.v1.legacyopenshift` and `roles.v1beta1.rbac.k8s.io` would seem to be a counterexample.  I see no reason to believe we won't collide more in the future.

Comment (smarterclayton) [inline pkg/cmd/util/clientcmd/factory.go]: At worst I can default exclude the cohabitators, I guess.

File: pkg/cmd/util/clientcmd/factory.go [hunk 6]
```diff
@@ -774,3 +776,43 @@ func (c *clientCache) ClientForVersion(version *unversioned.GroupVersion) (*clie
...
+			if strings.Contains(r.Name, "/") {
+				continue
+			}
+			// because discovery info doesn't tell us whether the object is virtual or not, perform a lookup
+			// by the kind for resource (which should be the canonical resource) and then verify that the reverse
+			// lookup (KindsFor) does not error.
+			if mapping, err := m.RESTMapping(unversioned.GroupKind{Group: gv.Group, Kind: r.Kind}, gv.Version); err == nil {
+				if _, err := m.KindsFor(mapping.GroupVersionKind.GroupVersion().WithResource(mapping.Resource)); err == nil {
+					gr := unversioned.GroupResource{Group: mapping.GroupVersionKind.Group, Resource: mapping.Resource}
+					set.Insert(gr.String())
```
Comment (deads2k) [inline pkg/cmd/util/clientcmd/factory.go]: I'm still not seeing why you don't just have `map[unversioned.GroupResource]bool` to keep the types correct.  I don't see a reason to convert through a string.

**Comment by smarterclayton** [comment]
Updated with your nit-picks.

**Comment by deads2k** [comment]
[merge]

**Comment by deads2k** [comment]
flake: https://github.com/openshift/origin/pull/9501

**Comment by smarterclayton** [comment]
[merge]

**Comment by smarterclayton** [comment]
Slow master brings all the flakes to the yard.
